pkg/storegateway/series_refs.go

```diff
@@ -741,19 +741,24 @@ func openBlockSeriesChunkRefsSetsIterator(
 	matchers []*labels.Matcher, // Series matchers.
 	shard *sharding.ShardSelector, // Shard selector.
 	seriesHasher seriesHasher,
-	skipChunks bool, // If true chunks are not loaded and minTime/maxTime are ignored.
+	strategy seriesIteratorStrategy,
 	minTime, maxTime int64, // Series must have data in this time range to be returned (ignored if skipChunks=true).
 	chunkRangesPerSeries int,
 	stats *safeQueryStats,
+	ps []storage.SeriesRef, // If this is not empty, these posting are used as it is without fetching new ones.
+	pendingMatchers []*labels.Matcher, // This is used in conjunction with 'ps'.
 	logger log.Logger,
-) (seriesChunkRefsSetIterator, error) {
+) (seriesChunkRefsSetIterator, []storage.SeriesRef, []*labels.Matcher, error) {
 	if batchSize <= 0 {
-		return nil, errors.New("set size must be a positive number")
+		return nil, nil, nil, errors.New("set size must be a positive number")
 	}
 
-	ps, pendingMatchers, err := indexr.ExpandedPostings(ctx, matchers, stats)
-	if err != nil {
-		return nil, errors.Wrap(err, "expanded matching postings")
+	if len(ps) == 0 {
+		var err error
+		ps, pendingMatchers, err = indexr.ExpandedPostings(ctx, matchers, stats)
+		if err != nil {
+			return nil, nil, nil, errors.Wrap(err, "expanded matching postings")
+		}
 	}
```

Comment:
If there were simply no postings the previous time (selectors matched 0 series), then this will trigger `ExpandedPostings` again for this evaluation.

Reply:
The easiest solution is to not execute the streaming chunks code if there were 0 series when streaming the series. I will do that.